x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -129,83 +161,112 @@ func (k Keeper) UpdateDelegationBalances(ctx sdk.Context, hostZone types.HostZon
 		}
 	}
 
-	if latestCompletionTime.IsZero() {
-		return nil, errorsmod.Wrapf(types.ErrInvalidPacketCompletionTime, "Invalid completion time (%s) from txMsg", latestCompletionTime.String())
+	if latestCompletionTime == 0 {
+		return 0, errorsmod.Wrapf(types.ErrInvalidPacketCompletionTime, "Invalid completion time 0 from txMsg")
 	}
-	return &latestCompletionTime, nil
+	return latestCompletionTime, nil
 }
 
-// UpdateHostZoneUnbondings does two things:
-//  1. Update the time of each hostZoneUnbonding on each epochUnbondingRecord
-//  2. Return the number of stTokens that need to be burned
-func (k Keeper) UpdateHostZoneUnbondings(
+// Updates the host zone unbonding records after a successful undelegation batch
+// The StTokensToBurn and the NativeTokensToUnbond amounts on the records are
+// decremented in a cascading fashion starting from the earliest record
+// The latest completion times is also set on each record if the time from the
+// batch is later than what's currently on the record
+func (k Keeper) UpdateHostZoneUnbondingsAfterUndelegation(
 	ctx sdk.Context,
-	latestCompletionTime time.Time,
 	chainId string,
-	undelegateCallback types.UndelegateCallback,
-) (stTokenBurnAmount sdkmath.Int, err error) {
-	stTokenBurnAmount = sdkmath.ZeroInt()
-	for _, epochNumber := range undelegateCallback.EpochUnbondingRecordIds {
-		epochUnbondingRecord, found := k.RecordsKeeper.GetEpochUnbondingRecord(ctx, epochNumber)
+	epochUnbondingRecordIds []uint64,
+	totalNativeTokensUnbonded sdkmath.Int,
+	unbondingTime uint64,
+) (totalStTokensToBurn sdkmath.Int, err error) {
+	// As we process the accounting changes, keep track of the stTokens that should be burned later
+	totalStTokensToBurn = sdkmath.ZeroInt()
+
+	// Loop each epoch unbonding record starting from the earliest
+	for _, epochNumber := range epochUnbondingRecordIds {
+		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochNumber, chainId)
 		if !found {
-			errMsg := fmt.Sprintf("Unable to find epoch unbonding record for epoch: %d", epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+			return totalStTokensToBurn, errorsmod.Wrapf(recordstypes.ErrHostUnbondingRecordNotFound,
+				"host zone unbonding not found for epoch %d and %s", epochNumber, chainId)
 		}
-		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochUnbondingRecord.EpochNumber, chainId)
-		if !found {
-			errMsg := fmt.Sprintf("Host zone unbonding not found (%s) in epoch unbonding record: %d", chainId, epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+
+		// Determine the amount to decrement from the record, flooring at 0
+		// Also decrement the total for th next loop
+		nativeTokensUnbonded := sdkmath.MinInt(hostZoneUnbonding.NativeTokensToUnbond, totalNativeTokensUnbonded)
```

Comment:
Is the `MinInt` just here for redundancy? Could `totalNativeTokensUnbonded` ever be larger than `hostZoneUnbonding.NativeTokensToUnbond`?

Reply:
It will actually quite often be greater than hostZoneUnbonding.NativeTokensToUnbond because there's multiple records!

E.g. Two EUR, each with 1000 native
Total would be 2000
In the first loop, `MinInt` would set `nativeTokensUnbonded` to `min(1000, 2000) = 1000` and then only decrement the record by 1000

Reply:
AH yes of course. Signed off on this then!